src/SubscribeBlockTracker.test.ts

```diff
@@ -0,0 +1,3221 @@
+                  response: {
+                    result: '0x0',
+                  },
+                },
+                {
+                  methodName: 'eth_unsubscribe',
+                  response: {
+                    result: true,
+                  },
+                },
+                {
+                  methodName: 'eth_blockNumber',
+                  response: {
+                    result: '0x1',
+                  },
+                },
+                {
+                  methodName: 'eth_subscribe',
+                  response: {
+                    result: '0x1',
+                  },
+                },
+                {
+                  methodName: 'eth_unsubscribe',
+                  response: {
+                    result: true,
+                  },
+                },
+              ],
+            },
+            blockTracker: blockTrackerOptions,
+          },
+          async ({ provider, blockTracker }) => {
+            const sendAsyncSpy = jest.spyOn(provider, 'sendAsync');
+            await blockTracker[methodToGetLatestBlock]();
+            // For PollingBlockTracker, there are possibly multiple
+            // `setTimeout`s in play at this point. For SubscribeBlockTracker
+            // that is not the case, as it does not poll, but there is no harm
+            // in doing this.
+            await setTimeoutRecorder.nextMatchingDuration(
+              blockTrackerOptions.blockResetDuration,
+            );
+            await blockTracker[methodToGetLatestBlock]();
+            const requestsForLatestBlock = sendAsyncSpy.mock.calls.filter(
+              (args) => {
+                return args[0].method === 'eth_blockNumber';
+              },
+            );
+            expect(requestsForLatestBlock).toHaveLength(2);
+          },
+        );
+      });
+
+      METHODS_TO_ADD_LISTENER.forEach((methodToAddListener) => {
+        it(`should not emit the "error" event (added via \`${methodToAddListener}\`) and should resolve with undefined if the request for the latest block number returns an error response`, async () => {
+          recordCallsToSetTimeout({ numAutomaticCalls: 1 });
+
+          await withSubscribeBlockTracker(
+            {
+              provider: {
+                stubs: [
+                  {
+                    methodName: 'eth_blockNumber',
+                    response: {
+                      error: 'boom',
+                    },
+                  },
+                  {
+                    methodName: 'eth_subscribe',
+                    response: {
+                      result: '0x64',
+                    },
+                  },
+                ],
+              },
+            },
+            async ({ provider, blockTracker }) => {
+              const promiseForCaughtError = new Promise<any>((resolve) => {
+                blockTracker[methodToAddListener]('error', resolve);
+              });
+              blockTracker.on('_started', () => {
```

Comment:
Hmm. I don't follow what this is doing

Reply:
Curious that this test used an `eth_subscribe` stub, but the following three did not

Reply:
>In this case, however, I do care, because I need to emit the "data" event with a subscription parameter that matches the id of the subscription created earlier.

OK I follow most of this, but this is where I got lost. Why do you need to emit the "data" event with a matching subscription ID in this case?

The test seems to be about using `on`/`addListener` to listen for the `error` event, specifically in the circumstance where `eth_blockNumber` returns an invalid response. I see that the error event never gets emitted in this case. I don't see what role this subscription plays. The test still seems to pass even after I comment it out.

Reply:
After looking at this again, you're right, stubbing `eth_subscribe` isn't needed here. Fixed in [1024248](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/10242483a5fdab9f0830df15a0fc2d178b33a640).